### DISCUSSION TITLE
Reading raw chunk files

### DISCUSSION THREAD

**Discussion by puneet336** [prometheus/prometheus#12076] [Q&A]
### Proposal

Hi ,
I have some old prometheus data , and as soon as i point the prometheus server to the old data (using storage.tsdb.path) , prometheus start fails and i see error messages which indicate failure of scrape, rule, notifier managers and i get opening storage failed error.

Is there any utility through which i can read the raw data chunks  (example - /data/prometheus/01GMFBC,…0/chunks/000001)?
Please do let me know if i can provide any further information from my end.

**Comment by roidelapluie** [comment]
Hi @puneet336,

Thank you for reaching out to us with your query. Yes, there is a utility called "promtool tsdb dump" that you can use to read the raw data. It will help you in dumping the samples from a TSDB.

You can use the below command to use the "promtool tsdb dump" tool:

```
promtool tsdb dump <flags> <db path>
```

Please note that you need to specify the path of the TSDB used by Prometheus in place of `<db path>`. Also, replace `<flags>` with any options that you need to provide.

Hope this helps!

**Reply by puneet336** [reply]
Hi @roidelapluie 
Thank you for your reply. 
**Summary** -  I attempted to use the 'promtool tsdb dump' command, but it was not successful for me. 

**Details -** 
Specifically, I am dealing with some corrupted data in the /database/prometheus.org directory that was generated using prometheus version 2.8.1 (branch: HEAD, revision: 4d60eb36dcbed725fcac5b27018574118f12fffb). My goal is to transfer this data to the /database/prometheus_puneettest directory. However, the version of prometheus that I have installed (2.8.1) does not support the 'tsdb dump' feature, so I downloaded version 2.37.6 which does support it. When I used promtool v2.37.6, this is what I found:

![image](https://user-images.githubusercontent.com/5935825/224425353-ac1b1f7d-fa5b-450a-a5b4-0f0b01b17162.png)

I tried recovering a slice of data (Tuesday, July 26, 2022 12:00:00 AM to Thursday, July 28, 2022 12:00:00 AM) from - 
`01GABDSGYHJ7R5W6ZN7CZD8DNY  2022-07-25 12:00:00 +0000 UTC  2022-08-13 06:00:00 +0000 UTC  450h0m0s     25363780613  215157077    945413       8GiB851MiB937KiB1010B`
as - 
![image](https://user-images.githubusercontent.com/5935825/224426391-2d5a5740-e13d-4b52-bfd8-97badc70aba3.png)

So it seems i need a existing database directory,  So i started a new prometheus instance 

![image](https://user-images.githubusercontent.com/5935825/224427198-2ce70001-bc93-444e-8409-78373d0e7efd.png)
I did not enabled any targets for scraping, as I aimed to load `only` historical data first.
<img width="392" alt="image" src="https://user-images.githubusercontent.com/5935825/224429269-59893057-db45-4d3d-b0d0-907bc71ea650.png">



and retried the same command. from new tab . Command ran fine  with exit/return code 0 
![image](https://user-images.githubusercontent.com/5935825/224427234-8fafa157-5489-4397-8f0d-d416ee12d41b.png)

But after inspecting the data in the /database/prometheus_puneettest directory, I did not observe the expected results.
![image](https://user-images.githubusercontent.com/5935825/224427566-e1e5b158-27db-44bf-8744-09e3d241d566.png)


Did i miss out on anything here?
Please do let me know if some more information is required from my  end here.

**Reply by puneet336** [reply]
Update:
I also have a separate directory (/database/prometheus) where the current prometheus instance is storing its data. However, the promtool was not able to retrieve the data from this source and store it in the /database/prometheus_puneettest directory. 



![image](https://user-images.githubusercontent.com/5935825/224430664-d2514ba8-02f5-4733-81a1-956e4b1ee3eb.png)

I am beginning to suspect that there may be an incompatibility between the data format of version 2.8.1 and promtool v2.37.6.   Reason of choosing v2.37.6 is that  the version of prometheus that I have installed (2.8.1) does not support the 'tsdb dump' feature, so I downloaded version 2.37.6 which does support it.

**Reply by roidelapluie** [reply]
When you have the message 'wal: no such file or directory', you can just create an empty wal directory.

**Reply by puneet336** [reply]
ok, since you had mentioned "`<db path>`" i was under impression that a prometheus database should exist.
I retried with just ` empty wal directory` but promtool did not work successfully for me.
![image](https://user-images.githubusercontent.com/5935825/224482722-54cf2d96-b26c-4b1a-ad4c-9c206cf52eb5.png)

Not sure what why exit status of promtool is reported 0 even though it was unable to dump the data.
Please do let me know if any further information is required from my end.
os - CentOS 7

**Comment by roidelapluie** [comment]
But you don't have blocks in that try2 directory

**Reply by puneet336** [reply]
ok, 
Summary  -I copied  over a "block" from the existing monitoring data directory  and tried the promtool tsdb dump again. Unfortunately, result is still the same.



Detail - 
Try1: 
As you had mentioned that we need to have a block , So started the prometheus again with  /database/prometheus_recovery_try2/ as data directory as -  
```
}~/MySoftwares/prometheus-2.37.6.linux-amd64/prometheus --web.listen-address="0.0.0.0:9999" --storage.tsdb.path=$PWD --config.file=$HOME/MySoftwares/prometheus-2.37.6.linux-amd64/prometheus.yml
```
and this time,  i have started the monitoring of localhost:9090 (from prometheus.yml) as well to get some data blocks. Here's the contents of try2 directory after starting monitoring ( 15 minutes) - 
![image](https://user-images.githubusercontent.com/5935825/224510457-2a36281f-f75d-406c-a36d-5ae20003dd3a.png)
So  promtool did not work.


Try2:
i manually copied 1 block directory - 
![image](https://user-images.githubusercontent.com/5935825/224510713-39739222-6ff9-4914-9415-22b311834c01.png)
now i can see some blocks , i attempted the tsdb command again _ 

![image](https://user-images.githubusercontent.com/5935825/224510806-c5c50214-7a93-4f42-9cfb-0022fc40fcc8.png)
But even after getting some blocks in the try 2 directory, i was unable to get the promtool to recover the data.

Please advice.

**Reply by puneet336** [reply]
Hi @roidelapluie , 
I tried creating a block by copying over some data files , but that did not work (try2 in the above reply).
Do you see issues with the way i am attempting to recover the data ?

Unable to understand that if promtool was unable to recover the data then why the exit status is 0.
